text/0000-overlapping_match.md

```diff
@@ -0,0 +1,183 @@
+matches then it skips checking any of the other branches and jumps to the end of the expression.
+
+Implementation:
+1. This can be implemented as if it was a list of `if` expressions. And a flag to check if any
+of the branches have been visited so as to not visit the `else`
+2. To cover the `else` case the location to jump to at the end after checking all the branches
+can be stored, initially set to the start of the `else` block but if it enters any of the
+branches then it is set to immediately after the `else` block.
+
+# How We Teach This
+[how-we-teach-this]: #how-we-teach-this
+
+This should be called `match/in` expressions since that is the combination of keywords that are
+used similar to `for/in` expressions. This idea would be best presented as a continuation of
+existing Rust patterns since it expands on the `match` expression.
+
+This proposal should be introduced to new users right after `match` expressions are taught. This
+is the best time to teach it since it appears as an extension of that syntax and the ideas that
+are used when using `match` expressions.
+
+Within the _Rust Book_ a section after the section on the `_` placeholder could be called
+_match/in Control Flow Operator Addition_. Within this section the syntax and differences would
+be outlined. These would most notable include the multiple branches can be executed. The reader
+should be able to understand by the end of this section that this allows for multiple branches
+to be executed but it still will check for exhaustiveness when able. He should also know that
+the branches are checked top first.
+
+An example that could be used within the section:
+
+You can turn this:
+```rust
+match cmp.compare(&array[left], &array[right]) {
+    Less => {
+        merged.push(array[left]);
+        left += 1;
+    },
+    Equal => {
+        merged.push(array[left]);
+        merged.push(array[right]);
+        left += 1;
+        right += 1;
+    },
+    Greater => {
+        merged.push(array[right]);
+        right += 1;
+    }
+}
+```
+into
+```rust
+match cmp.compare(&array[left], &array[right]) in {
+    Less | Equal => {
+        merged.push(array[left]);
+        left += 1;
+    },
+    Greater | Equal => {
+        merged.push(array[right]);
+        right += 1;
+    }
+}
+```
+
+Another example is an implementation of fizzbuzz:
+
+```rust
+for x in 1...100 {
+    let mut res = String::from("");
+    if x % 5 == 0 {
+        res += "fizz";
+    }
+    if x % 7 == 0 {
+        res += "buzz";
+    }
+    if res.len() == 0 {
+        res = x.to_string();
+    }
+    println!("{}", res);
+}
+```
+into
+```rust
+for x in 1...100 {
+    match x in {
+        _ if x % 5 == 0 => print!("fizz"),
+        _ if x % 7 == 0 => print!("buzz")
+    } else {
+        print!("{}", x);
+    }
+    println!("");
+}
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This should not be done because it increases the size of language and might not be used by
+everyone.
+
```

Comment:
I'd like to see some discussion of the readability of inclusive top-down multiple-branch `match` constructs.